scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using a "hash_id" managed by the clients. It uniquely identifies the resource's manifest.
+
+ScannerV4 will replace ScannerV2 as the container image scanner in Central. A new image integration for ScannerV4 will be created and replace ScannerV2's integration (type "clarify"), activated by the feature flag: `ROX_SCANNER_V4_ENABLED`. The "clarify" type will continue to be used for orchestrator and node scanning.
+
+ScannerV4 gRPC endpoints will use GRPC status codes to communicate the return status in case of non-successful responses.
+
+ScannerV4 will also provide endpoints for health and observability checks, which are not detailed in this ADR.
+
+### ScannerV4 API Endpoints
+
+The endpoints are versioned at `v4` to align with Clair:
+
+1. `scanner.v4.Indexer/CreateIndex`: Create a manifest of a specified resource and create an index or re-index. Idempotent while creation is being executed. Synchronous and tied to the client's request. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19)
+2. `scanner.v4.Indexer/GetIndex`: Retrieve or check index existence. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19).
+3. `scanner.v4.Matcher/GetVulnerabilities`: Get vulnerabilities for a given resource's manifest. Returns [`VulnerabilityReport`](https://github.com/quay/claircore/blob/v1.4.18/vulnerabilityreport.go#L7).
+4. `scanner.v4.Matcher/GetVulnerabilityMetadata`: Get information on vulnerability metadata, e.g., last update timestamp.
+
+Example of manifest's `hash_id` usage to create index reports for container images:
+
+```
+message ContainerImageLocator {
+    string url      = 1;
+    string username = 2;
+    string password = 3;
+}
+
+message CreateIndexRequest {
+    string hash_id = 1;
+    oneof resource_locator {
```

Comment:
Locators can be either an image or another type of locator, such as a Node. These types are heterogeneous. It's important to note that we may want to use multiple APIs to index different locators, considering their distinct promises, availability, and latency. It could be worthwhile to consider using "CreateImageIndex," "CreateNodeIndex," and so on.